EWS is not adding merging-blocked label on PRs

JonWBedard opened the pull request:
#### a2ee6d8fbe53342bd0aa22969c6008716dc2e27f
<pre>
EWS is not adding merging-blocked label on PRs
<a href="https://bugs.webkit.org/show_bug.cgi?id=252244">https://bugs.webkit.org/show_bug.cgi?id=252244</a>
rdar://105452651

Reviewed by Aakash Jain.

_is_hash_outdated is an asynchronous function, we should treat it as such.

* Tools/CISupport/ews-build/steps.py:
(GitHubMixin.should_send_email_for_pr): _is_hash_outdated is asynchronous, so this function is too.
(BlockPullRequest.run): Treat _is_hash_outdated as asynchronous.
(AnalyzeCompileWebKitResults.send_email_for_new_build_failure): Treat should_send_email_for_pr as asynchronous.
(AnalyzeLayoutTestsResults.send_email_for_new_test_failures): Ditto.

Canonical link: <a href="https://commits.webkit.org/261420@main">https://commits.webkit.org/261420@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f955b4cf53b7e9646ad987b8b07593bfa407ba78

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/6/builds/111586 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/20721 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/90/builds/190 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/3345 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/120343 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/115646 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/22077 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/11803 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/87/builds/3345 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/19/builds/117351 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/22077 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/90/builds/190 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/104374 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/22077 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/90/builds/190 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/104374 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/13206 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/90/builds/190 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/104374 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/13701 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/11803 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| [  ~~🧪 services~~](https://ews-build.webkit.org/#/builders/20/builds/110358 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/19145 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/90/builds/190 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/15683 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4334 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

aj062 reviewed (approved):
r=me assuming this is tested

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -3824,10 +3834,10 @@ def send_email_for_new_test_failures(self, test_names, exceed_failure_limit=Fals
             if not change_author:
-                self._addToLog('stderr', 'Unable to determine email address for {} from metadata/contributors.json. Skipping sending email.'.format(self.getProperty('owners', [])))
+                yield self._addToLog('stderr', 'Unable to determine email address for {} from metadata/contributors.json. Skipping sending email.'.format(self.getProperty('owners', [])))
                 return
```
Do we need defer.returnValue(None) here?

JonWBedard replied:
See above comment, no, that's the default.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2837,8 +2842,10 @@ def send_email_for_new_build_failure(self):
+                if not should_send_email:
+                    return
             if not patch_id and not (pr_number and sha):
                 self._addToLog('stderr', 'Unrecognized change type')
```
this also needs `yield`

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2837,8 +2842,10 @@ def send_email_for_new_build_failure(self):
+            if pr_number:
+                should_send_email = yield self.should_send_email_for_pr(pr_number, self.getProperty('repository'))
+                if not should_send_email:
+                    return
```
Do we need defer.returnValue(None) in all the returns inside this method (now that it has `@defer.inlineCallbacks`)?

JonWBedard replied:
The way I like to think about this is that `defer.inlineCallbacks` turns every `return` into a `break`. We don't generally need to do `defer.returnValue(None)` because that's the default of `defer.inlineCallbacks` (similar to how these three statements are identical in Python)
```
def f_a():
    pass

def f_b():
    return

def f_c():
    return None
```

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2829,6 +2833,7 @@ def filter_logs_containing_error(self, logs, searchString='rror:', max_num_lines
                 filtered_logs.append(line)
         return '\n'.join(filtered_logs[-max_num_lines:])
 
+    @defer.inlineCallbacks
```
since `send_email_for_new_build_failure` now has `@defer.inlineCallbacks`, should it's invocation inside AnalyzeCompileWebKitResults.analyzeResults() also have `yield`? (or is that not necessary?)

JonWBedard replied:
Pedantically, yes. And I intend to do that in the coming days.

Practically, though, no. `send_email_for_new_build_failure` is basically operating as a sub-routine instead of a function. Since its caller doesn't care about the result, its caller doesn't need to wait for the result. Now, this means that `send_email_for_new_build_failure` can actually finish _after_ it's caller does, which is a bit weird and not great, but it shouldn't break anything.

The reason I didn't include modifications to `AnalyzeCompileWebKitResults.analyzeResults` in this change is because those changes are higher risk, and I was trying to fix the failure to apply the `merging-blocked` label as quickly as possible.

aj062 replied:
ok, sounds good.

JonWBedard committed:
EWS is not adding merging-blocked label on PRs
https://bugs.webkit.org/show_bug.cgi?id=252244
rdar://105452651

Reviewed by Aakash Jain.

_is_hash_outdated is an asynchronous function, we should treat it as such.

* Tools/CISupport/ews-build/steps.py:
(GitHubMixin.should_send_email_for_pr): _is_hash_outdated is asynchronous, so this function is too.
(BlockPullRequest.run): Treat _is_hash_outdated as asynchronous.
(AnalyzeCompileWebKitResults.send_email_for_new_build_failure): Treat should_send_email_for_pr as asynchronous.
(AnalyzeLayoutTestsResults.send_email_for_new_test_failures): Ditto.

Canonical link: https://commits.webkit.org/261420@main
